### PR TITLE
WIP: Add basic vagrant provisioining

### PR DESCRIPTION
**Pull request by alayek**
- [X] Vagrant provisioning
- [X] git
- [X] Node
- [X] Yarn & NVM
- [X] MongoDB
- [ ] Cypress
- [ ] Test on Windows

## How to test
- ` git checkout` the PR
- install VirtualBox and Vagrant (`brew cask install virtualbox vagrant`)
- run `vagrant up` in the root directory of the project, which has the `Vagrantfile`. This would take about 6 m (on OSX 10) to provision the Ubuntu box. On Windows, this may be closer to 10 mins.
- log into the vagrant box with `vagrant ssh`
- check versions of `git`, `node`, `mongo`
- set Mongo Path. Will add it to provisioning later.
- go to `/vagrant` with `cd /vagrant`. This directory is shared by default with your project. Vagrant does it by default.
- Run `yarn setup`. This would be running inside the Ubuntu machine. Might give some error related to Cypress install.

### CODE REVIEW

**Comment by kentcdodds** [comment]
This is super! I'll give it a shot tomorrow probably (I'm traveling right now). But this is so awesome. Thank you so much for your help :D

**Comment by kentcdodds** [comment]
Thank you SO MUCH for this! It's awesome. So I got everything installed perfectly. It worked great. And when I ran `yarn setup` things worked really well except for loading the database with fake data. I got `Error: EACCESS: permission denied, mkdir './mongo-db'`:

<details>

<summary>yarn setup</summary>

```
vagrant@vagrant:/vagrant$ yarn setup
yarn setup v0.21.3
$ node ./scripts/verify && node ./scripts/install && node ./scripts/load-database && npm start validate 
🎉  Congrats! Your system is setup properly
You should be good to install and run things.
📦  Installing dependencies via yarn --force --ignore-platform --ignore-engines
🔑  starting install in /vagrant
yarn install v0.21.3
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
warning "eslint-import-resolver-babel-module@3.0.0" has unmet peer dependency "babel-core@^6.0.0".
[4/4] Rebuilding all packages...
success Saved lockfile.
Done in 60.34s.
🎉  finished installing dependencies in "/vagrant"
🔑  starting install in /vagrant/api
yarn install v0.21.3
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Rebuilding all packages...
success Saved lockfile.
Done in 31.76s.
🎉  finished installing dependencies in "/vagrant/api"
🔑  starting install in /vagrant/client
yarn install v0.21.3
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Rebuilding all packages...
success Saved lockfile.
$ nps postinstall
nps executing: echo "removing client copy of eslint (will use root repo's version)" && node node_modules/rimraf/bin.js node_modules/eslint
removing client copy of eslint (will use root repo's version)
Done in 31.57s.
🎉  finished installing dependencies in "/vagrant/client"
👍  all dependencies installed
stopping mongod with `mongo admin --eval "db.shutdownServer()" --port 27017` (in case it is running)
MongoDB shell version v3.4.2
connecting to: mongodb://127.0.0.1:27017/admin
2017-03-25T13:00:37.780+0000 W NETWORK  [thread1] Failed to connect to 127.0.0.1:27017, in(checking socket for error after poll), reason: Connection refused
2017-03-25T13:00:37.805+0000 E QUERY    [thread1] Error: couldn't connect to server 127.0.0.1:27017, connection attempt failed :
connect@src/mongo/shell/mongo.js:237:13
@(connect):1:6
exception: connect failed
/vagrant/scripts/load-database.js:16
      throw err
      ^

Error: EACCES: permission denied, mkdir '/.mongo-db'
    at Error (native)
error Command failed with exit code 1.
```

</details>

I tried:

```
vagrant@vagrant:/vagrant$ chown $(whoami) .mongo-db
```

And that command succeeded but the `node ./scripts/load-database.js` script still failed for the same reason. Any ideas?

**Comment by kentcdodds** [comment]
I ran a few more scripts to make sure other things were working. Pretty much everything's working _except_ starting and connecting to mongo. Also When running `npm start e2e.run.cypress` (which should start cypress in headless mode), it successfully installs a fresh copy of cypress, but then I get this error:

```
 -Error- 
Could not start Cypress headlessly. Your CI provider must support XVFB.
```

I think @edm00se may have some experience with this...

And when I try to open the cypress GUI:

```
npm start e2e.dev.cypress
```

I get the same error, which is interesting because that script runs:

```
./node_modules/.bin/cypress open
```

which should actually open the GUI, not headless mode.

And running the GUI is the main reason we're doing this in the first place ;-)

Thanks again for your work on this, I hope we can get over these hurdles!

**Comment by alayek** [comment]
@kentcdodds yes, there are a few issues, and I will fix them now.

I have a question about Cypress. Are we trying to run Cypress as if it's a server running inside the VM, and some port forwarding would make it accessible to the browser on host system (OSX / Windows)?

**Comment by kentcdodds** [comment]
Maybe [this](http://stackoverflow.com/a/33138627/971592) or [this](http://stackoverflow.com/a/20024671/971592) could help...

**Comment by kentcdodds** [comment]
I have to run, but I was able to successfully start an Ubuntu GUI, but I couldn't find the project source files. It was weird. At first there was all this stuff in `/vagrant`, now that folder is empty. 🤔

**Comment by edm00se** [comment]
Ooh, that looks to be coming along nicely. If I get time later, I may clone things and poke around.

**Comment by alayek** [comment]
@kentcdodds I have a workaround for this database load error.

Note that `/vagrant` is a top-level directory, residing right alongside other directories like `boot`, `etc`, `sbin` etc. In this case, if you notice the error message, `mongoPath` is being resolved to `/.mongo-db`, i.e, another top level directory.

Clearly, only `root` can access this location. I have a workaround for this:

Set the `MONGO_PATH` environment variable to be `/home/vagrant/.mongo-db`, or `/vagrant/.mongo-db` in `~/.bashrc`. Typically, mongo likes to go with `/data/db`, but that has a different issue of having to run the `mongod` command with `sudo` privileges.

**TL;DR**: this problem can be easily solved if we set the `MONGO_PATH` environment variable, pointing to a directory the `vagrant` user can modify.

---

I would caution against going with GUI. Vagrant is an OS on top of another existing OS. Even systems with 8 GB RAM would start to feel slow. With GUI and a running database, that problem would be exacerbated to whole different level. The Yarn link step can cause issues too, because it eats up too much CPU time. 

Sure, we can control the amount of memory allocated, but this is a potential limitation to keep in mind.

There is another part - the workflow. I understand that this is a workshop, so you would take control of the machine and demonstrate. But if we are making a case for people at large to be able to use this on their own time, we might want to think whether we would ask them to execute things in a Vagrant machine.

The user at large doesn't necessarily care how this runs, as long they can run Cypress and Mongo in their OS.

Docker is a pretty lightweight and competing solution for this. And both OSX and Windows 10 have good Docker support. 

As an end user, I simply want to run Mongo DB and Cypress while I progress through the workshop; inside a Docker / Vagrant. To that end, commands like `docker exec [args]` or `vagrant exec [args]` can be pretty helpful. You execute once to start your DB / Cypress, and continue with the lesson on your host system, in your favorite text editor / IDE.

**Comment by edm00se** [comment]
So, the intention I believe @kentcdodds had in going with the VM was chiefly to have a reproducible envoronment in which to make use of Cypress's GUI. If it was to just run headlessly, we _could_ do so with the Docker version from PR #16. The GUI with Cypress is the goal, most modern dev machines should have enough system resources to run a Linux desktop environment, although to make it more user friendly, using an ultralight desktop environment like LXDE might be preferable.

**Comment by alayek** [comment]
Ok, I will add the GUI.

**Comment by kentcdodds** [comment]
Thanks so much @alayek! We no longer need this because Cypress now supports windows :)
